### PR TITLE
foundry.md: draft startPrank and stopPrank

### PR DESCRIPTION
**Pull request by anvacaru**
part of: https://github.com/runtimeverification/foundry/issues/12

### CODE REVIEW

File: foundry.md [hunk 1]
```diff
@@ -120,6 +131,46 @@ First we have some helpers in K which can:
...
 ```
 
+Prank
+-----
+```k
+    rule [call.startPrank]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <caller> CL => #asAccount(#range(LM, ARGSTART +Int 4, 32)) </caller>
+         <origin> OG </origin>
```
Comment (ehildenb) [inline foundry.md:141]: The formatting here is strange.

Comment (ehildenb) [reply foundry.md:141]: Let's at least make sure the cells line up.

File: foundry.md [hunk 2]
```diff
@@ -97,7 +109,6 @@ First we have some helpers in K which can:
          </account>
 ```
 ```k
-
     rule [call.deal]:
```
Comment (ehildenb) [inline foundry.md:112]: Here as well.

File: foundry.md [hunk 3]
```diff
@@ -120,6 +131,46 @@ First we have some helpers in K which can:
...
+    rule [call.startPrankWithOrigin]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 => 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <caller> CL => #asAccount(#range(LM, ARGSTART +Int 4, 32)) </caller>
+         <origin> OG => #asAccount(#range(LM, ARGSTART +Int 36, 32)) </origin>
+         <prevCaller> _ => CL </prevCaller>
+         <prevOrigin> _ => OG </prevOrigin>
+         <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 1169514616 // selector ( "startPrank(addr)" )
```
Comment (ehildenb) [inline foundry.md:158]: These two cheatcodes have the same signature, but different selector integers, which doesn't make sense to me.

File: foundry.md [hunk 4]
```diff
@@ -145,6 +157,70 @@ This rule then takes the account using `#asWord(#range(LM, ARGSTART +Int 4, 32)`
...
+    rule [call.stopPrank]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH 0 0 ~> 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <caller> _ => CL </caller>
+         <origin> _ => OG </origin>
+         <prevCaller> CL </prevCaller>
+         <prevOrigin> OG </prevOrigin>
+         <localMem> LM </localMem>
+      requires CHEAT_ADDR ==Int #address(FoundryCheat)
+       andBool CL =/=K .Account andBool OG =/=K .Account
```
Comment (lisandrasilva) [inline foundry.md:219]: Why do we need this? Is this a possible scenario?

Comment (anvacaru) [reply foundry.md:219]: I was thinking about corner case scenarios in which `stopPrank()` is called before a `startPrank`, but I have to double check how foundry actually handles those tests.

Comment (ehildenb) [reply foundry.md:219]: @anvacaru please write a test which calls stopPrank before startPrank, so we can make sure we handle it similarly to Foundry.

File: foundry.md [hunk 5]
```diff
@@ -145,6 +157,69 @@ This rule then takes the account using `#asWord(#range(LM, ARGSTART +Int 4, 32)`
          </account>
 ```
 
+Pranks
+------
+
+#### `startPrank` - Sets `msg.sender` for all subsequent calls until `stopPrank` is called.
```
Comment (anvacaru) [inline foundry.md:163]:
```suggestion
#### `startPrank` - Sets `msg.sender` and `tx.origin` for all subsequent calls until `stopPrank` is called.
```

File: foundry.md (hunk 4 above)
Comment (anvacaru) [reply foundry.md]: Done. Looks like Foundry accepts tests in which `stopPrank` is called before `startPrank`. I've updated the rules accordingly.

**Review by lisandrasilva — APPROVED** [review]

File: tests/foundry/exclude [hunk 6]
```diff
@@ -68,6 +68,9 @@ NonceTest.testNonce
 OwnerUpOnlyTest.testFailIncrementAsNotOwner
 OwnerUpOnlyTest.testIncrementAsNotOwner
 OwnerUpOnlyTest.testIncrementAsOwner
+PlainPrankTest.test_startPrankWithOrigin_true
+PlainPrankTest.test_stopPrank_true2
```
Comment (RaoulSchaffranek) [inline tests/foundry/exclude:72]: Why are these tests excluded? Shouldn't they pass?

File: tests/foundry/exclude [hunk 7]
```diff
@@ -68,6 +68,9 @@ NonceTest.testNonce
...
+PlainPrankTest.testFail_startPrank_false
```
Comment (RaoulSchaffranek) [inline tests/foundry/exclude:73]: Hmmm... We should find a way to exercise this test and confirm that it fails instead of just disabling it.

File: tests/foundry/test/PlainPrankTest.t.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function test_stopPrank_true2() public {
+        AdditionalToken token = new AdditionalToken();
+        vm.startPrank(address(token));
+        token.incrementCount();
+        assert(token.count() == 1);
+    }
+
+    function testFail_startPrank_false() public {
```
Comment (RaoulSchaffranek) [inline tests/foundry/test/PlainPrankTest.t.sol:52]: I don't understand what is tested here and in which line the test case is expected to fail.

File: foundry.md [hunk 9]
```diff
@@ -145,6 +157,69 @@ This rule then takes the account using `#asWord(#range(LM, ARGSTART +Int 4, 32)`
...
+The `#if` statements in cells `<id>` and `<origin>` are used in order to avoid the rewriting of the cells to `.Account` if there is no prank active.
+
+```k
+    rule [call.stopPrank]:
+        <k> CALL _ CHEAT_ADDR 0 ARGSTART _ARGWIDTH _ 0 => 1 ~> #push ... </k>
+         <output> _ => .ByteArray </output>
+         <id> CL => #if OLD_CL =/=K .Account #then OLD_CL #else CL #fi </id>
+         <origin> OG => #if OLD_OG =/=K .Account #then OLD_OG #else OG #fi </origin>
+         <prevId> OLD_CL => .Account </prevId>
+         <prevOrigin> OLD_OG => .Account </prevOrigin>
```
Comment (RaoulSchaffranek) [inline foundry.md:216]: Is this identical to what Foundry is doing? I can imagine that calls to `startPrank` stack, however, the public documentation is not clear about that.

File: tests/foundry/exclude (hunk 6 above)
Comment (JuanCoRo) [reply tests/foundry/exclude:72]: I think so. Maybe they are there to be removed from the file when the last failing test passes. I'll try them out tomorrow.

File: tests/foundry/exclude (hunk 7 above)
Comment (JuanCoRo) [reply tests/foundry/exclude:73]: As far as I know, we haven't yet implemented the mechanisms for `testFail`, which I believe is the reason for the last test to be excluded. I will double-check that.

File: tests/foundry/test/PlainPrankTest.t.sol (hunk 8 above)
Comment (JuanCoRo) [reply tests/foundry/test/PlainPrankTest.t.sol:52]: My understanding is that the test will fail the second time `startPrank` is called since `stopPrank` has not been called before. But I don't know if this is the sole purpose of this test. In any case, I'll isolate this behavior into a new one.

File: foundry.md [hunk 10]
```diff
@@ -76,6 +82,12 @@ Foundry Cheat Codes
 module FOUNDRY-CHEAT-CODES
     imports EVM
     imports FOUNDRY-ACCOUNTS
+
+    configuration
+    <cheatcodes>
+        <prevId> .Account </prevId>
+        <prevOrigin> .Account </prevOrigin>
+    </cheatcodes>
```
Comment (JuanCoRo) [inline foundry.md:90]: These cells will add to those in #1349 and could be conceptually confusing. Maybe it's better to create a cell that represents the [`Prank`](https://github.com/foundry-rs/foundry/blob/master/evm/src/executor/inspector/cheatcodes/env.rs#L36) struct. Should come in handy for adding the `prank` cheatcode in the future.
Also, notice that the `depth` is missing, which would cause a `startPrank` to be active for external calls of more depth.
